### ISSUE TITLE
Tool-independent encryption and licensing

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#1282]
**Reported by ahaumer on 23 Sep 2013 09:05 UTC**
Tool-independent encryption and licensing would support small Modelica consultancy companies in selling libraries, no matter which tool the customer chooses. The annotations are already defined in the Modelica specification. We're only lacking a tool for encryption and decrypttion. 

We would suggest to send out an MA call to implement:
* a stand-alone tool that allows to encrypt a single Modelica-file (.mo) along with a textual license file (.mo), both to an encrypted Modelica-file (.moe and .mo-lic). Example of a license file:

```
package MyLicenseFile
  annotation(Authorization(
    libraryKey="xxxxxx-xxxxxx-xxxxxx-xxxxxx",
    licensor="Organization X",
    license="{License(licensee="Organization Y", 
                      id={"mac:yyyyyyyyyyyy"},
                      expirationDate="yyyy-mm-dd")}));
end MyLicenseFile;
```
* a decryption library (dll) for Windows and Linux that allows tool vendors to implement the same decryption of the both above mentioned files (.moe and .mo-lic) with low effort.
The mechanism in the decryption library looks for the mo-lic file, decrypts it and  whether it is valid, and if yes, decrypts the Modelica file (.moe) in allocated memory, enabling the tool to read the Modelica code from memory. 
Of course, the tool is responsible to obey all the rules for encrypted files as specified in the Modelica specification.

Final remark:
We don't want to discuss endlessly for a perfect solution, but to start with an easy-to-implement solution as soon as possible.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1282

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 23 Sep 2013 13:34 UTC**
I cannot see how this can be made reasonable safe in the proposal above.

The issues are:
**DLLs are bad for protecting contents:**
It is fairly easy to use debugging tools to get all information for DLL-calls. Additionally if we have the source for the decryption library one could make a debug-version that outputs all the information to a file as well.
That issue can be avoided by just providing source/object file insted.

**Decryption software has export restrictions**
This is really complicated, and we would need to involve lawyers for this.

**It gives a false sense of safety**
Based on the code above it would be trivial to build the single-file decrypt standalone program (it would be natural as part of code development/debugging); we cannot rely on no-one doing that. 

There ways to be safe even if the code is known are:
* Secret key (or half-secret with public-key crypt) - but does not help here, since the user need some access to it - without full access.
* Tool-specific public key. That means the one providing the library still need to know for which tool it is encrypted.
* There are some theoretical ideas for solving the underlying problem, but not clear how practical.

Basically this is similar to making an open-source library that allows playing encrypted movies/music, but prevents ripping.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 15 Nov 2013 13:30 UTC**
80th Design Meeting is long over ... closing that milestone and still open tickets set back to \*undecided\*.

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 20 Jun 2014 10:58 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 10 Jun 2015 06:35 UTC**
There has been some further discussion of the encryption topic at milestone:Design86, minutes: [source:ModelicaDesign/trunk/MeetingMinutesMaterial/min86_2015_Waterloo/Slides-and-Documents/EncryptionDiscussion_2015_06_09.docx EncryptionDiscussion_2015_06_09.docx]

**Comment by modelica-trac-importer** [comment]
**Comment by ahaumer on 19 Dec 2015 22:13 UTC**
Also see #1868 Suggestion for a standard for encryption, licensing and packeting of libraries

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 31 Mar 2016 14:03 UTC**
Replying to [comment:5 ahaumer]:
> Also see #1868 Suggestion for a standard for encryption, licensing and packeting of libraries

Closing this one as duplicate of #1868.
